Add tests for reports

1pretz1 opened the pull request:
This PR adds tests for some (of the many) rake tasks. I focused mainly on the Publishing API rake tasks as they were the least tested and high value. There has also been some refactoring to try make navigating and adding new publishing api tasks a bit easier.

---

Trello:
https://trello.com/c/LCR0Am73/2541-5-enable-continuous-deployment-for-whitehall

ChrisBAshton commented on test/unit/reports/organisation_attachments_report_test.rb:
```diff
@@ -0,0 +1,42 @@
+    assert_equal Reports::OrganisationAttachmentsReport::CSV_HEADERS, CSV.read(path)[0]
+    assert_equal 1, CSV.read(path, headers: true).count
+
+    File.delete(path) if File.exist?(path)
```
We can probably remove the conditional here.

ChrisBAshton commented on test/unit/reports/organisation_attachments_report_test.rb:
```diff
@@ -0,0 +1,42 @@
+    organisation = publication.organisations.last
+
+    path = nil
+    Timecop.freeze do
```
Could we move all of the test contents inside this block, and remove the slightly hacky `path = nil` initialisation? Same applies to the other tests.

1pretz1 replied:
Sure thing

ChrisBAshton commented on test/unit/reports/published_attachments_report_test.rb:
```diff
@@ -0,0 +1,35 @@
+  end
+
+  test "returns blank report if there are no eligible file attachments" do
+    create(:publication, :with_file_attachment)
```
So a publication can't have a file attachment? 🤔 Or it can, but we just don't want it included in the report?

1pretz1 replied:
It can but it won't get returned in the report unless the document is published. I've matched the factories document types (both now detailed guides) which should make the difference in test setup more apparent.

ChrisBAshton replied:
Makes sense, thanks 👍

ChrisBAshton commented on app/models/redirect_route.rb:
```diff
@@ -0,0 +1,11 @@
+      },
+    ]
+  end
+end
```
Nice 👍

ChrisBAshton commented on lib/tasks/publishing_api.rake:
```diff
@@ -100,13 +61,7 @@ namespace :publishing_api do
-        content_id: "36620cf9-00b3-4d51-afab-e05e457061e3",
-      },
-    ].each do |route|
+    RedirectRoute.all.each do |route|
```
Yep, much nicer to keep config away from logic.

ChrisBAshton commented on test/unit/tasks/publishing_api_test.rb:
```diff
@@ -0,0 +1,384 @@
+      test "republishes record by slug" do
+        tasks.each do |task|
+          record = create(task[:model].underscore) # rubocop:disable Rails/SaveBang
+          model = task[:model].constantize
```
The more conventional approach to this kind of DRYing up of tests is to use [shared example groups](https://relishapp.com/rspec/rspec-core/v/2-0/docs/example-groups/shared-example-group).

1pretz1 replied:
I wish this was RSpec world, unfortunately we're currently faking RSpec features in Minitest which doesn't include shared examples, I don't think.

ChrisBAshton replied:
Ah yes, I'd missed that this is Minitest.

ChrisBAshton commented on test/unit/tasks/publishing_api_test.rb:
```diff
@@ -0,0 +1,384 @@
+          record = create(task[:model].underscore) # rubocop:disable Rails/SaveBang
+          model = task[:model].constantize
+
+          model.any_instance.expects(:publish_to_publishing_api).at_least_once
```
The `at_least_once` threw me slightly - do we know how many times it is called? Should it just be once?

1pretz1 replied:
Yes it should be just once - `at _least_once` was copied over from the test below. Turns out that `publish_to_publishing_api` was being called more than once due to the Role factory creating an Organisation (and attempting to publish to pub api) on initialisation. Altered the test config slightly to mitigate this. Good spot!!

ChrisBAshton commented on test/unit/tasks/publishing_api_test.rb:
```diff
@@ -0,0 +1,384 @@
+  extend Minitest::Spec::DSL
+
+  teardown do
+    task.reenable if defined?(task) # without this, calling `invoke` does nothing after first test
```
It's a shame this doesn't get applied to the `republish` set of tests, though it could if we defined a `let(:task)` in there. Can we do that?

1pretz1 replied:
Agreed so I've tried to refactor using `let`, with not much success. I'm open to suggestions though 😄

ChrisBAshton replied:
I've had a go here: https://github.com/alphagov/whitehall/commit/fe6d249da76630a37862f1d8a33bf442e7352d5d
It does lead to _slightly_ more lines of code, but I think this is worth it for the increased readability.

The rest of the PR looks great @1pretz1, but think it's worth making this one last tweak before merge ✅

1pretz1 replied:
Sure thing, I've split them into their own separate tests. I was in two minds when writing them as it did seem like a compromise on consistency when trying to DRY up the code.

ChrisBAshton commented on test/unit/tasks/publishing_api_test.rb:
```diff
@@ -0,0 +1,384 @@
+      describe "#document_type" do
+        let(:task) { Rake::Task["publishing_api:bulk_republish:document_type"] }
+
+        test "republishes all documents from document type" do
```
```suggestion
        test "republishes all documents of the specified document type" do
```

ChrisBAshton commented on test/unit/tasks/publishing_api_test.rb:
```diff
@@ -0,0 +1,384 @@
+          task.invoke(org.slug)
+        end
+
+        test "Ignores documents owned by other organisation" do
```
```suggestion
        test "Ignores documents owned by other organisations" do
```

ChrisBAshton commented on test/unit/tasks/publishing_api_test.rb:
```diff
@@ -0,0 +1,384 @@
+
+          task.invoke(filename)
+
+          File.delete("lib/tasks/#{filename}.csv") if File.exist?("lib/tasks/#{filename}.csv")
```
I think we can remove this conditional too.

ChrisBAshton commented on test/unit/tasks/publishing_api_test.rb:
```diff
@@ -0,0 +1,384 @@
+          CSV.open("lib/tasks/#{filename}.csv", "wb") do |csv|
+            csv << %w[content_id]
+            csv << [edition.content_id]
+          end
```
Might be nicer to use a string for this, something like:

```ruby
File.write("lib/tasks/#{filename}.csv", "content_id\n#{edition.content_id}")
```

1pretz1 replied:
Nice suggestion, done!

ChrisBAshton commented on lib/tasks/publishing_api.rake:
```diff
@@ -84,6 +84,12 @@ namespace :publishing_api do
+  desc "Republish all organisations"
+  task republish_all_organisations: :environment do
+    Organisation.find_each(&:publish_to_publishing_api)
+  end
```
Nice 👍

ChrisBAshton reviewed (commented):
Really good job bringing the test coverage up, Peter 👍 I have a few questions and suggestions below.

1pretz1 commented:
Thanks for the review @ChrisBAshton! Think I've addressed everything.

I was getting some strange behaviour when I was running the rake tests locally (tasks running twice, when only invoked once) and have narrowed it down to the tasks themselves being initialised more than once. I've updated this in the last commit.

ChrisBAshton reviewed (approved):
Great work, @1pretz1 👏
